### PR TITLE
DEV: Initial Commit

### PR DESCRIPTION
**Pull request by janzenisaac**
This PR adds the core functionality of the plugin. 

# Overview
There is a gap in Discourse Core when it comes to tracking a user's session... especially for anon users.

This plugin adds a `page_visits` table to Discourse Core that then commits a user's 

- user_id (if user is authenticated)
- ip_address
- user_agent
- topic_id (if visiting a topic route)
- post_ids (if visiting a topic route, we record every post_id that was read, even if it has been read before)
- visit_time (stored in ms)

for each visit session. This is available for both anonymous and authenticated users. A session is ended, and data is written to the DB on (one of) two conditions: 

1. They change pages (or routes). 
2. They trigger a `visiblitychange` event. 

While changing routes is a simple pattern to follow and catch, `visiblitychange` is not. Following [documentation](https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/) referred to by MDN, there is no _catch all_ solution for all desktop / mobile browsers, for tracking when a browser session has ended. The best and most widely accepted solution is to place your analytic callbacks in a 

```js
if (document.visibilityState === "hidden")
```

block, which this PR introduces.

### CODE REVIEW

File: assets/javascripts/discourse/api-initializers/page-visits.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+function resetPageVisitData() {
+  pageVisitData = {};
+}
+
+async function createPageVisitRecord(data, postIds, timer) {
+  try {
+    await ajax("/page_visits", {
+      type: "POST",
```
Comment (janzenisaac) [inline assets/javascripts/discourse/api-initializers/page-visits.js:93]: I am getting the error 

```
Routing Error

No route matches [POST] "/page_visits"
```

Looking at rails info routes, it exists as I would expect: 

Screenshot 2024-06-26 at 12 58 27 PM

- I tried including the controller file within the `plugin.rb`, eg `require_relative "app/controllers/discourse_page_visits/page_visits_controller"`

Not sure what the issue is here.

Comment (CvX) [reply assets/javascripts/discourse/api-initializers/page-visits.js:93]: `.json` perhaps?

Comment (janzenisaac) [reply assets/javascripts/discourse/api-initializers/page-visits.js:93]: @CvX sadly no, tried that as well.

Comment (CvX) [reply assets/javascripts/discourse/api-initializers/page-visits.js:93]: Ah, I see now. You're mounting the engine like this:

```rb
mount ::DiscoursePageVisits::Engine, at: "discourse-page-visits"
```

so the path is `/discourse-page-visits/page_visits`, right?

File: assets/javascripts/discourse/api-initializers/page-visits.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    // screenTrack index is 1-based
+    const postId = postStream.stream[index - 1];
+    if (postId && !viewedPostIds.includes(postId)) {
+      viewedPostIds.push(postId);
+    }
+  });
+}
+
+async function createPageVisitRecord(data, postIds, time) {
+  await ajax("/page_visits.json", {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/api-initializers/page-visits.js:66]: Ideally the rails convention would be `/page-visits.json`

File: assets/javascripts/discourse/api-initializers/page-visits.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  });
+}
+
+async function createPageVisitRecord(data, postIds, time) {
+  await ajax("/page-visits.json", {
+    type: "POST",
+    data: {
+      user_id: data.userId,
+      full_url: data.fullUrl,
+      user_agent: data.userAgent,
```
Comment (markvanlan) [inline assets/javascripts/discourse/api-initializers/page-visits.js:71]: I think you can remove userAgent from the frontend here since we capture in Ruby

**Review by markvanlan — APPROVED** [review]
Looks great!

**Pull request merged**